pennylane/qnode.py

```diff
@@ -410,6 +373,163 @@ def construct(self, args, kwargs=None):
+
+        for param_idx, gate_param_tuple in variable_ops_sorted:
+            # iterate over all parameters
+            for op_idx, _ in gate_param_tuple:
+                # get all dependents of the existing parameter
+                sub = set(nx.dag.topological_sort(G.subgraph(nx.dag.ancestors(G, op_idx)).copy()))
+
+                # check if any of the dependents are in the
+                # existing layer
+                if set(layer_ops[layer]["ops"]) & sub:
+                    # operation depends on previous layer,
+                    # start a new layer count
+                    layer += 1
+
+                # store the parameters and ops indices for the layer
+                layer_ops.setdefault(layer, {"ops": [], "pidx": []})
+                layer_ops[layer]["ops"].append(op_idx)
+                layer_ops[layer]["pidx"].append(param_idx)
+
+        # iterate through each layer
+        for _, v in layer_ops.items():
+            # for the layer, get the ops and parameter indices
+            ops = v["ops"]
+            param_idx = v["pidx"]
+
+            # get the ops in this layer
+            curr_ops = [op for n, op in G.nodes(data="op") if n in ops]
+
+            # get all ancestor operations
+            ancestors = set()
+            for o in ops:
+                subG = G.subgraph(nx.dag.ancestors(G, o))
+                ancestors |= set(subG.nodes(data="op"))
+
+            ancestors = sorted(ancestors, key=lambda x: x[0])
+            queue = [op for _, op in ancestors if op not in curr_ops]
+
+            obs = []
+            first_order = []
+            second_order = []
+            V = None
+            scale = []
+
+            # for each operator, get the generator
+            # and convert it to a variance
+            for n, op in enumerate(curr_ops):
+                gen, s = op.generator
+                w = op.wires
+
+                if gen is None:
+                    raise QuantumFunctionError("Can't generate metric tensor, operation {}"
+                                               "has no defined generator".format(op))
+
+                # get the observable corresponding
+                # to the generator of the current operation
+                if isinstance(gen, np.ndarray):
+                    # generator is a Hermitian matrix
+                    variance = pennylane.var(pennylane.Hermitian(gen, w, do_queue=False))
+
+                    if not diag_approx:
+                        first_order.append((n, expand(gen, w, self.num_wires)))
+
+                elif issubclass(gen, pennylane.operation.Observable):
+                    # generator is an existing PennyLane operation
+                    variance = pennylane.var(gen(w, do_queue=False))
+
+                    if not diag_approx:
+                        if issubclass(gen, pennylane.ops.PauliX):
+                            mat = np.array([[0, 1], [1, 0]])
```

Comment:
We could make the matrices realizing the operators properties of `Operation` and `Observable`.

Reply:
That's a good idea, I think we should :+1:

As this is quite a major change, should we make an issue for it?

Reply:
Probably a good idea -- but I mean in the CV gates you also have the `_heisenberg_rep`, so it's actually not that far fetched.